### PR TITLE
Read SigLens Memory from cgroups

### PR DESCRIPTION
**Pull request by Macbeth98**
# Description
- Read memory from cgroups by detecting cgroup version and falling back to host memory.

# Testing
- Tested on my local docker with Kubernetes enabled and the it worked fine with cgroup v2
- Tested by creating a gcp instance with docker image set to cgroupv1 and verified that the memory is being read properly

# Checklist:

- [x] I have self-reviewed this PR.
- [x] I have removed all print-debugging and commented-out code that should not be merged.
- [x] I have added sufficient comments in my code, particularly in hard-to-understand areas.
- [x] I have formatted the code, if applicable. For Go, I have run `goimports -w .`.

### CODE REVIEW

File: pkg/config/config.go [hunk 1]
```diff
@@ -100,23 +112,134 @@ func init() {
...
+		memoryFilePaths.UsagePaths = []string{cgroupV2UsageMemoryPath}
+	} else {
+		memoryFilePaths.MaxPaths = []string{cgroupV1MaxMemoryPath}
+		memoryFilePaths.UsagePaths = []string{cgroupV1UsageMemoryPath}
+	}
+
+	// Add Kubernetes-specific paths only when inside Kubernetes
+	memoryFilePaths.MaxPaths = append(memoryFilePaths.MaxPaths,
+		"/sys/fs/cgroup/kubepods/memory.max",                 // Kubernetes cgroup v2
+		"/sys/fs/cgroup/kubepods.slice/memory.max",           // Kubernetes cgroup v2 with systemd
```
[the remark this answers is not in the document]
Comment (Macbeth98) [reply pkg/config/config.go:166]: I wanted to add them regardless of kubernetes env, but I added the change to add only when inside a kubernetes env.

File: pkg/config/common/common.go [hunk 2]
```diff
@@ -84,6 +84,8 @@ type DatabaseConfig struct {
 }
 
 type MemoryConfig struct {
+	MaxMemoryAllowedToUseInBytes uint64 `yaml:"maxMemoryAllowedToUseInBytes"`
```
[the remark this answers is not in the document]
Comment (Macbeth98) [reply pkg/config/common/common.go:87]: If it is `0`, it is ignored. 
I have added the test cases.
But I want to keep it simple as `uint64` only. I added a comment mentioning that a `0` value will be ignored.

File: pkg/config/config.go [hunk 3]
```diff
@@ -100,23 +112,134 @@ func init() {
...
+
+	detectCgroupVersion()
+	initMemoryPaths()
+}
+
+func detectCgroupVersion() {
+	data, err := os.ReadFile("/proc/mounts")
+	if err != nil {
+		log.Warnf("detectCgroupVersion: Error reading /proc/mounts: %v", err)
+	} else {
```
[the remark this answers is not in the document]
Comment (Macbeth98) [reply pkg/config/config.go]: So this is assuming that the docker containers will be on linux. I have added the code to skip the function if the OS is not `linux`.

File: pkg/config/config.go [hunk 4]
```diff
@@ -84,8 +84,7 @@ const (
 type cgroupVersion uint8
 
 const (
-	noCgroupVersion cgroupVersion = iota
-	cgroupVersion1
+	cgroupVersion1 cgroupVersion = iota
```
Comment (AndrewHess) [inline pkg/config/config.go:87]: In `detectCgroupVersion` you return 0 to indicate there's no cgroup. But that's actually `cgroupVersion1`, so maybe you need `cgroupVersion1 cgroupVersion = iota + 1`

File: pkg/config/config.go [hunk 5]
```diff
@@ -216,7 +247,7 @@ func detectCgroupPath(basePath string) string {
 }
 
 func getContainerMemory(memoryValueType memoryValueType) (uint64, error) {
-	if cgroupVersionDetected == noCgroupVersion {
+	if cgroupVersionDetected == 0 {
```
Comment (AndrewHess) [inline pkg/config/config.go:250]: Hmm I actually think the way you had it before was more readable. So I'd suggest changing it back

Sorry for my bad advice to remove `noCgroupVersion`

File: pkg/config/config_test.go [hunk 6]
```diff
@@ -241,13 +246,14 @@ func Test_ExtractConfigData(t *testing.T) {
...
 				EnableSortIndex:             utils.DefaultValue(false),
 				MemoryConfig: common.MemoryConfig{
-					MaxUsagePercent: 80,
-					LowMemoryMode:   utils.DefaultValue(false),
-					SearchPercent:   DEFAULT_SEG_SEARCH_MEM_PERCENT,
-					CMIPercent:      DEFAULT_ROTATED_CMI_MEM_PERCENT,
-					MetadataPercent: DEFAULT_METADATA_MEM_PERCENT,
-					MetricsPercent:  DEFAULT_METRICS_MEM_PERCENT,
-					BytesPerQuery:   DEFAULT_BYTES_PER_QUERY,
+					MaxMemoryAllowedToUseInBytes: uint64(math.Min(10_000_000_000, float64(memory.TotalMemory()))),
```
Comment (AndrewHess) [inline pkg/config/config_test.go:249]: Nice

**Review by AndrewHess — COMMENTED** [review]
Looks pretty good; just a few minor things

File: pkg/config/config.go (hunk 4 above)
Comment (Macbeth98) [reply pkg/config/config.go:87]: Ahh yes!

File: pkg/config/config.go (hunk 5 above)
Comment (Macbeth98) [reply pkg/config/config.go]: No worries!

**Review by AndrewHess — APPROVED** [review]

**Pull request merged**